firedrake/assemble.py

```diff
@@ -2103,21 +2085,45 @@ def index_function_spaces(form, indices):
         else:
             raise AssertionError
 
-    @staticmethod
-    def index_tensor(tensor, form, indices, diagonal):
-        """Return the PyOP2 data structure tied to ``tensor``, indexed
-        if necessary.
-        """
-        rank = len(form.arguments())
-        is_indexed = any(i is not None for i in indices)
 
-        if rank == 0:
-            return tensor
-        elif rank == 1 or rank == 2 and diagonal:
+def _as_pyop2_tensor(tensor, indices=None):
+    """Cast a Firedrake tensor into a PyOP2 data structure, optionally indexing it."""
+    is_indexed = indices and any(index is not None for index in indices)
+
+    if isinstance(tensor, op2.Global):
+        assert not is_indexed
+        return tensor
+
+    if isinstance(tensor, (firedrake.Function, firedrake.Cofunction)):
+        if is_indexed:
             i, = indices
-            return tensor.dat[i] if is_indexed else tensor.dat
-        elif rank == 2:
+            return tensor.dat[i]
+        else:
+            return tensor.dat
+    else:
+        assert isinstance(tensor, firedrake.Matrix)
+        if is_indexed:
             i, j = indices
-            return tensor.M[i, j] if is_indexed else tensor.M
+            return tensor.M[i, j]
         else:
-            raise AssertionError
+            return tensor.M
```

Comment:
> This free function checks types and return things according to the types. The best OOP practice I think is to implement this as a method of ParloopFormAssembler (see _as_pyop2_type), so I think we should just modify _as_pyop2_type or add some similar method to ParloopFormAssembler.
Attaching such methods to ParloopFormAssembler makes sense as ParloopFormAssembler is the one which walks over blocks and supply correctly index tensors to parloops.

I actually reckon that the best way to do this would be with a `singledispatch` function. I think that the process of extracting arguments for parloops is more generic than just for assembly (e.g. we might one day want this for interpolation) so attaching it to an assembly-specific class isn't right.

> In this regard, we should probably just pass an indexed tensor to ParloopBuilder.build() in the first place as this will limit the scope of ParloopBuilder to a block of the matrix that is being assembled instead of the entire matrix, but this might also require moving need_unrolling and collect_lgmaps to ParloopFormAssembler. This is beyond the scope of this PR.

Yes something like this makes sense. I would be tempted to try and avoid using the builder pattern in future though because I think it gets a bit messy.

Reply:
Doesn't that just mean that we should refactor interpolation.py so that we can reuse things from assembler? If it turns out in the future that extracting arguments for parloops is useful in other places, we should probably just come up with an abstraction to allow for `tensor.pyop2data.as_parloop_arg`. I do not think a free function or a singledispatch would be the right thing to introduce. So I still suggest that you save refactorisation for a future PR and focus here on integrating the contents of `_as_parloop_type`.

Reply:
I've checked and making this not be a free function is a bit messy because `as_pyop2_type` is called inside both `ParloopFormAssembler` and `ParloopBuilder`.

I agree that the ideal behaviour is for PyOP2 to do the right thing, but I am unwilling to invest time in developing PyOP2 when I am actively working on its replacement.

I'd be happy to discuss this tomorrow in person if that would be useful (I will be in the office).

Reply:
Can you restore changes around `ParloopBuilder` and just do something like:
```
class ExplicitMatrixAssembler():

    ...

    @staticmethod
    def _as_pyop2_tensor(tensor, indices=None):
        is_indexed = indices and any(index is not None for index in indices)
        if is_indexed:
            i, j = indices
            tensor_ = tensor.M[i, j]
        else:
            tensor_ = tensor
        if tensor_.handle.getType() == "python":
            mat_context = pyop2_tensor.handle.getPythonContext()
            if isinstance(mat_context, _GlobalMatPayload):
                return mat_context.global_
            elif isinstance(mat_context, _DatMatPayload):
                return mat_context.dat
            else:
                raise ValueError
        else:
            return tensor_
```
do similar things in `ZeroFormAssembler` and `OneFormAssembler`, and use these methods at appropriate place in `ParloopAssembler`?

We should note that we currently need to index the output tensor in `_as_parloop_arg_output` because we pass unindexed tensor to `ParloopBuilder`. As mentioned earlier, we should probably just pass indexed tensor to `ParloopBuilder`, in which case indexing in `_as_parloop_arg_output` will become unnecessary.

Reply:
Done.